language/diem-vm/parallel_executor/src/scheduler.rs

```diff
@@ -0,0 +1,104 @@
+use crossbeam_utils::CachePadded;
+use std::{
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, RwLock,
+    },
+};
+
+const FLAG_NOT_EXECUTED: usize = 0;
+const FLAG_EXECUTED: usize = 1;
+
+pub struct Scheduler {
+    // Shared index (version) of the next txn to be executed from the original transaction sequence.
+    execution_marker: AtomicUsize,
+    // Shared number of txns to execute: updated before executing a block or when an error or
+    // reconfiguration leads to early stopping (at that transaction version).
+    stop_at_version: AtomicUsize,
+
+    txn_buffer: SegQueue<usize>, // shared queue of list of dependency-resolved transactions.
+    txn_dependency: Vec<CachePadded<Arc<RwLock<Option<Vec<usize>>>>>>, // version -> txns that depend on it.
+    txn_status: Vec<AtomicUsize>, // version -> execution status.
+}
+
+impl Scheduler {
+    pub fn new(num_txns: usize) -> Self {
+        Self {
+            execution_marker: AtomicUsize::new(0), stop_at_version: AtomicUsize::new(num_txns),
+            txn_buffer: SegQueue::new(),
+            txn_dependency: (0..num_txns)
+                .map(|_| CachePadded::new(Arc::new(RwLock::new(Some(Vec::new())))))
+                .collect(),
+            txn_status: (0..num_txns)
+                .map(|_| AtomicUsize::new(FLAG_NOT_EXECUTED))
+                .collect(),
+        }
+    }
+
+
+    // Return the next txn id for the thread to execute: first fetch from the shared queue that
+    // stores dependency-resolved txns, then fetch from the original ordered txn sequence.
+    // Return Some(id) if found the next transaction, else return None.
+    pub fn next_txn_to_execute(&self) -> Option<usize> {
+        // Fetch txn from txn_buffer
+        match self.txn_buffer.pop() {
+            Some(version) => Some(version),
+            None => {
+                // Fetch the first non-executed txn from the original transaction list
+                let next_to_execute = self.execution_marker.fetch_add(1, Ordering::Relaxed);
+                if next_to_execute < self.num_txn_to_execute() {
+                    Some(next_to_execute)
+                } else {
+                    // Everything executed at least once - validation will take care of rest.
+                    None
+                }
+            }
+        }
+    }
+
+    // Invoked when txn depends on another txn, adds version to the dependency list the other txn.
+    // Return true if successful, otherwise dependency resolved in the meantime, return false.
+    pub fn add_dependency(&self, version: usize, dep_version: usize) -> bool {
+        // Could pre-check that the txn isn't in executed state, but shouldn't matter much since
+        // the caller usually has just observed the read dependency (so not executed state).
+
+        // txn_dependency is initialized for all versions, so unwrap() is safe.
+        let mut stored_deps = self.txn_dependency[version].write().unwrap();
```

Comment:
I had a typo in refactoring, sorry - just found it in our prototype. here it has to be dep_version.